src/Analyzer.Reports/SarifReportWriter.cs

```diff
@@ -0,0 +1,188 @@
+                    {
+                        Id = evaluation.RuleId,
+                        // Name = evaluation.RuleId, TBD issue #198
+                        FullDescription = new MultiformatMessageString { Text = AppendPeriod(evaluation.RuleDescription) },
+                        Help = new MultiformatMessageString { Text = AppendPeriod(evaluation.Recommendation) },
+                        HelpUri = hasUri ? uri : null,
+                        MessageStrings = new Dictionary<string, MultiformatMessageString>()
+                        {
+                            { "default", new MultiformatMessageString { Text = AppendPeriod(evaluation.RuleDescription) } }
+                        },
+                        DefaultConfiguration = new ReportingConfiguration { Level = GetLevelFromEvaluation(evaluation) }
+                    });
+            }
+            return rulesDictionary[evaluation.RuleId];
+        }
+
+        private void ExtractResult(IEvaluation rootEvaluation, IEvaluation evaluation, string filePath)
+        {
+            foreach (var result in evaluation.Results.Where(r => !r.Passed))
+            {
+                this.sarifResults.Add(new Result
+                {
+                    RuleId = rootEvaluation.RuleId,
+                    Level = GetLevelFromEvaluation(rootEvaluation),
+                    Message = new Message { Id = "default" }, // should be customized message for each result 
+                    Locations = new[]
+                    {
+                        new Location
+                        {
+                            PhysicalLocation = new PhysicalLocation
+                            {
+                                ArtifactLocation = new ArtifactLocation
+                                {
+                                    Uri = new Uri(
+                                    UriHelper.MakeValidUri(
+                                        filePath.Replace(this.rootPath, string.Empty, StringComparison.OrdinalIgnoreCase)),
+                                    UriKind.Relative),
+                                    UriBaseId = UriBaseIdString,
+                                },
+                                Region = new Region { StartLine = result.LineNumber },
+                            },
+                        },
+                    }
+                });
+            }
+
+            foreach (var eval in evaluation.Evaluations.Where(e => !e.Passed))
+            {
+                this.ExtractResult(rootEvaluation, eval, filePath);
+            }
+        }
+
+        private FailureLevel GetLevelFromEvaluation(IEvaluation evaluation)
+        {
+            // The rule severity definition work item: https://github.com/Azure/template-analyzer/issues/177
+            return evaluation.Passed ? FailureLevel.Note : FailureLevel.Error;
+        }
+
+        private void PersistReport()
+        {
+            using Stream outputTextStream = this.reportFile.Create();
```

Comment:
I recommend creating the file in the constructor instead.  Since this is only called during disposal, if file creation fails, it could be very difficult to debug (from cryptic exception messages) and the time it took to scan the template(s) will be lost because results couldn't be saved.

Reply:
updated as suggested.

Reply:
also refactor the SarifReportWriter to persist results to SARIF file once a template file analyzed. so if analyze multiple files some of them failed, it still generates results for the succeeded. Please review.